File: binary-tree-level-order-traversal/solution102.go
// https://leetcode.com/problems/binary-tree-level-order-traversal/

package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// T: O(n), S: O(n)
func levelOrder(root *TreeNode) [][]int {
	var result [][]int
	if root == nil {
		return result
	}
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		levelSize := len(queue)
		var currLevel []int
		for i := 0; i < levelSize; i++ {
			currNode := queue[0]
			queue = queue[1:]
			currLevel = append(currLevel, currNode.Val)
			if currNode.Left != nil {
				queue = append(queue, currNode.Left)
			}
			if currNode.Right != nil {
				queue = append(queue, currNode.Right)
			}
		}
		result = append(result, currLevel)
	}
	return result
}

func main() {
	root1 := &TreeNode{Val: 3}
	root1.Left = &TreeNode{Val: 9}
	root1.Right = &TreeNode{Val: 20}
	root1.Right.Left = &TreeNode{Val: 15}
	root1.Right.Right = &TreeNode{Val: 7}
	fmt.Println(levelOrder(root1)) // [[3],[9,20],[15,7]]

	root2 := &TreeNode{Val: 1}
	fmt.Println(levelOrder(root2)) // [[1]]

	var root3 *TreeNode = nil
	fmt.Println(levelOrder(root3)) // []
}

File: binary-tree-level-order-traversal/Solution102.java
// https://leetcode.com/problems/binary-tree-level-order-traversal/

import java.util.*;

class Solution102 {

    // T: O(n), S: O(n)
    public List<List<Integer>> levelOrder(TreeNode root) {
        var result = new ArrayList<List<Integer>>();
        if (root == null) {
            return result;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            var currLevel = new ArrayList<Integer>();
            for (int i = 0; i < levelSize; i++) {
                var currNode = queue.poll();
                currLevel.add(currNode.val);
                if (currNode.left != null) {
                    queue.add(currNode.left);
                }
                if (currNode.right != null) {
                    queue.add(currNode.right);
                }
            }
            result.add(currLevel);
        }

        return result;
    }

    public static void main(String[] args) {
        var sol = new Solution102();

        var root1 = new TreeNode(3);
        root1.left = new TreeNode(9);
        root1.right = new TreeNode(20);
        root1.right.left = new TreeNode(15);
        root1.right.right = new TreeNode(7);
        System.out.println(sol.levelOrder(root1)); // [[3],[9,20],[15,7]]

        var root2 = new TreeNode(1);
        System.out.println(sol.levelOrder(root2)); // [[1]]

        TreeNode root3 = null;
        System.out.println(sol.levelOrder(root3)); // []
    }

}
